src/java/org/apache/cassandra/service/reads/trackwarnings/TrackWarningsSnapshot.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.service.reads.trackwarnings;
+
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.google.common.collect.ImmutableSet;
+
+import org.apache.cassandra.locator.InetAddressAndPort;
+
+public class TrackWarningsSnapshot
+{
+    private static final TrackWarningsSnapshot EMPTY = new TrackWarningsSnapshot(Warnings.EMPTY, Warnings.EMPTY, Warnings.EMPTY);
+
+    public final Warnings tombstones, localReadSize, rowIndexTooLarge;
+
+    private TrackWarningsSnapshot(Warnings tombstones, Warnings localReadSize, Warnings rowIndexTooLarge)
+    {
+        this.tombstones = tombstones;
+        this.localReadSize = localReadSize;
+        this.rowIndexTooLarge = rowIndexTooLarge;
+    }
+
+    public static TrackWarningsSnapshot create(Warnings tombstones, Warnings localReadSize, Warnings rowIndexTooLarge)
+    {
+        if (tombstones == localReadSize && tombstones == rowIndexTooLarge && tombstones == Warnings.EMPTY)
+            return EMPTY;
+        return new TrackWarningsSnapshot(tombstones, localReadSize, rowIndexTooLarge);
+    }
+
+    public static TrackWarningsSnapshot merge(TrackWarningsSnapshot... values)
```

Comment:
A couple unit level tests to exercise the merging logic might make sense, although I do think I see the de-duping case now being hit in one of the in-JVM tests...

Reply:
>  I see the de-duping case now being hit in one of the in-JVM tests...

Where?  I checked circle and the tests are green

Reply:
created unit tests